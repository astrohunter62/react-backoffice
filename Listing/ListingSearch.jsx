import React from 'react'
import PropTypes from 'prop-types'

import ListingSearchBranch from './ListingSearchBranch'

const withListing = Component => class extends React.Component {
  static propTypes = {
    onFilter: PropTypes.func.isRequired,
  }

  constructor() {
    super()

    this.handleClick = this.handleClick.bind(this)
    this.handleFilter = this.handleFilter.bind(this)
  }

  state = {
    open: false,
  }

  handleClick() {
    this.setState({
      open: !this.state.open,
    })

    this.props.onFilter(undefined)
  }

  handleFilter(event) {
    const { value } = event.target

    if (this.state.open) {
      this.props.onFilter(value)
    } else {
      this.props.onFilter(undefined)
    }
  }

  render() {
    return (
      <Component
        {...this.props}
        {...this.state}
        onClick={this.handleClick}
        onFilter={this.handleFilter}
      />
    )
  }
}

export default withListing(ListingSearchBranch)
