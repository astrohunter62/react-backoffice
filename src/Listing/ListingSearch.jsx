import React from 'react'
import PropTypes from 'prop-types'

import ListingSearchBranch from './ListingSearchBranch'

const withListing = Component => class extends React.Component {
  static propTypes = {
    onFilter: PropTypes.func.isRequired,
  }

  constructor() {
    super()

    this.handleClick = this.handleClick.bind(this)
    this.handleFilter = this.handleFilter.bind(this)
    this.getSearchRef = this.getSearchRef.bind(this)
  }

  state = {
    open: false,
  }

  getSearchRef(node) {
    this.searchRef = node
  }

  handleClick() {
    const { open } = this.state
    const { onFilter } = this.props
    const newOpen = !open

    if (this.searchRef && newOpen) {
      this.searchRef.focus()
    }

    this.setState({
      open: newOpen,
    })

    onFilter(undefined)
  }

  handleFilter(event) {
    const { open } = this.state
    const { onFilter } = this.props
    const { value } = event.target

    if (open) {
      onFilter(value)
    } else {
      onFilter(undefined)
    }
  }

  render() {
    return (
      <Component
        {...this.props}
        {...this.state}
        onClick={this.handleClick}
        onFilter={this.handleFilter}
        getSearchRef={this.getSearchRef}
      />
    )
  }
}

export default withListing(ListingSearchBranch)
