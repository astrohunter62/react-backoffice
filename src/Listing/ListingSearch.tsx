import React, { FunctionComponent, useEffect, useRef, useState } from "react";
import classNames from "classnames";
import {
  TextField,
  Tooltip,
  IconButton,
  makeStyles,
  Theme,
} from "@material-ui/core";
import SearchIcon from "@material-ui/icons/Search";

const useStyles = makeStyles((theme: Theme) => ({
  root: {
    position: "relative",
    marginRight: theme.spacing(),
  },
  field: {
    position: "absolute",
    right: "100%",
    width: 0,
    marginTop: theme.spacing(),
    transition: "width 0.25s",
  },
  fieldActive: {
    width: theme.spacing(30),
  },
}));

type Props = {
  isOpen?: boolean;
  placeholder?: string;
  onFilter?: (...args: any[]) => any;
};

const ListingSearch: FunctionComponent<Props> = ({
  isOpen: isOpenProp,
  placeholder,
  onFilter,
}) => {
  const searchRef = useRef<HTMLInputElement>();
  const [isOpen, setIsOpen] = useState(isOpenProp);
  const classes = useStyles();

  useEffect(() => {
    if (isOpen) {
      searchRef.current?.focus();
    }
  }, [isOpen]);

  const handleFilter = (event: any) => {
    if (!onFilter) {
      return;
    }

    const { value } = event.target;

    onFilter(isOpen ? value : undefined);
  };

  const handleClick = () => {
    setIsOpen(!isOpen);

    if (onFilter) {
      onFilter(undefined);
    }
  };

  return (
    <div className={classes.root}>
      <TextField
        type="search"
        placeholder={placeholder}
        className={classNames(classes.field, {
          [classes.fieldActive]: isOpen,
        })}
        onChange={handleFilter}
        inputRef={searchRef}
      />

      <Tooltip title="Search">
        <IconButton onClick={handleClick}>
          <SearchIcon />
        </IconButton>
      </Tooltip>
    </div>
  );
};

export default ListingSearch;
