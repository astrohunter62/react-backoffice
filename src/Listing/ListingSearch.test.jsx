import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { IconButton } from '@material-ui/core'
import ListingSearch from './ListingSearch'

Enzyme.configure({ adapter: new Adapter() })

it('renders correctly', () => {
  const tree = shallow((
    <ListingSearch
      onClick={() => { }}
      onFilter={() => { }}
    />
  ))

  expect(tree).toMatchSnapshot()
})

it('renders correctly when open', () => {
  const tree = shallow((
    <ListingSearch
      open
      onClick={() => { }}
      onFilter={() => { }}
    />
  ))

  expect(tree).toMatchSnapshot()
})

it('opens if click is triggered', () => {
  const listingSearch = shallow((
    <ListingSearch
      open
      onClick={() => { }}
      onFilter={() => { }}
    />
  ))

  expect(listingSearch.instance().state.open).toEqual(false)

  listingSearch.find(IconButton).simulate('click')

  expect(listingSearch.instance().state.open).toEqual(true)
})

it('filters if open and value changes', () => {
  const onFilter = jest.fn()
  const listingSearch = shallow((
    <ListingSearch
      open
      onClick={() => { }}
      onFilter={onFilter}
    />
  ))

  listingSearch.find('input').simulate('change', { value: 'a' })

  expect(onFilter).toHaveBeenCalled()
})

it('does not filter if not open and value changes', () => {
  const onFilter = jest.fn()
  const listingSearch = shallow((
    <ListingSearch
      open
      onClick={() => { }}
      onFilter={onFilter}
    />
  ))

  expect(listingSearch.instance().state.open).toEqual(false)
  listingSearch.find(IconButton).simulate('click')

  expect(listingSearch.instance().state.open).toEqual(true)

  listingSearch.find('input').simulate('change', { value: 'a' })

  expect(onFilter).toHaveBeenCalled()
})

it('focuses search field on click', () => {
  const mockObject = {
    focus: jest.fn(),
  }

  const listingSearch = shallow((
    <ListingSearch
      onClick={() => { }}
      onFilter={() => {}}
    />
  ))

  listingSearch.instance().searchRef = mockObject

  listingSearch.find(IconButton).simulate('click')

  expect(listingSearch.state().open).toBe(true)

  expect(mockObject.focus).toHaveBeenCalled()
})
