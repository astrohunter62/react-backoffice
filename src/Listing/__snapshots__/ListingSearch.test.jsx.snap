// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="ListingSearchBranch-root-1"
>
  <div
    className="MuiFormControl-root-4 ListingSearchBranch-field-2"
  >
    <div
      className="MuiInput-root-8 MuiInput-formControl-9 MuiInput-underline-12"
    >
      <input
        aria-invalid={false}
        aria-required={false}
        className="MuiInput-input-16 MuiInput-inputType-19 MuiInput-inputTypeSearch-20"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Filter"
        type="search"
      />
    </div>
  </div>
  Array [
    "",
    <button
      className="MuiButtonBase-root-35 MuiIconButton-root-29"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiIconButton-label-34"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-38"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <g>
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </g>
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root-44"
      />
    </button>,
    <div
      className="MuiTooltip-popper-21"
      style={
        Object {
          "left": 0,
          "opacity": 0,
          "pointerEvents": "none",
          "position": "absolute",
          "top": 0,
        }
      }
    >
      <div
        aria-hidden={true}
        className="MuiTooltip-tooltip-23 MuiTooltip-tooltipPlacementBottom-28"
        role="tooltip"
      >
        Search
      </div>
    </div>,
  ]
</div>
`;

exports[`renders correctly when open 1`] = `
<div
  className="ListingSearchBranch-root-1"
>
  <div
    className="MuiFormControl-root-4 ListingSearchBranch-field-2"
  >
    <div
      className="MuiInput-root-8 MuiInput-formControl-9 MuiInput-underline-12"
    >
      <input
        aria-invalid={false}
        aria-required={false}
        className="MuiInput-input-16 MuiInput-inputType-19 MuiInput-inputTypeSearch-20"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Filter"
        type="search"
      />
    </div>
  </div>
  Array [
    "",
    <button
      className="MuiButtonBase-root-35 MuiIconButton-root-29"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiIconButton-label-34"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-38"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <g>
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </g>
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root-44"
      />
    </button>,
    <div
      className="MuiTooltip-popper-21"
      style={
        Object {
          "left": 0,
          "opacity": 0,
          "pointerEvents": "none",
          "position": "absolute",
          "top": 0,
        }
      }
    >
      <div
        aria-hidden={true}
        className="MuiTooltip-tooltip-23 MuiTooltip-tooltipPlacementBottom-28"
        role="tooltip"
      >
        Search
      </div>
    </div>,
  ]
</div>
`;
